game-core/src/main/java/games/strategy/triplea/ai/tree/BattleStep.java

```diff
@@ -0,0 +1,1216 @@
+    }
+    return returnFireAgainstAttackingSubs;
+  }
+
+  /*
+  // Taken from MustFightBattle::returnFireAgainstDefendingSubs
+  private MustFightBattle.ReturnFire returnFireAgainstDefendingSubs(
+  final Collection<Unit> attackingUnits, final Collection<Unit> defendingUnits) {
+    final boolean attackingSubsSneakAttack =
+        defendingUnits.stream().noneMatch(Matches.unitIsDestroyer());
+    final boolean defendingSubsSneakAttack =
+    defendingSubsSneakAttackAndNoAttackingDestroyers(attackingUnits);
+    final MustFightBattle.ReturnFire returnFireAgainstDefendingSubs;
+    if (!defendingSubsSneakAttack) {
+      returnFireAgainstDefendingSubs = MustFightBattle.ReturnFire.ALL;
+    } else if (attackingSubsSneakAttack || Properties.getWW2V2(parameters.getData())) {
+      returnFireAgainstDefendingSubs = MustFightBattle.ReturnFire.SUBS;
+    } else {
+      returnFireAgainstDefendingSubs = MustFightBattle.ReturnFire.NONE;
+    }
+    return returnFireAgainstDefendingSubs;
+  }
+   */
+
+  // Taken from MustFightBattle::defendingSubsSneakAttackAndNoAttackingDestroyers
+  private boolean defendingSubsSneakAttackAndNoAttackingDestroyers(
+      final Collection<Unit> attackingUnits) {
+    return attackingUnits.stream().noneMatch(Matches.unitIsDestroyer())
+        && defendingSubsSneakAttack();
+  }
+
+  // Taken from MustFightBattle::defendingSubsSneakAttack
+  private boolean defendingSubsSneakAttack() {
+    return Properties.getWW2V2(parameters.getData())
+        || Properties.getDefendingSubsSneakAttack(parameters.getData());
+  }
+
+  @Override
+  public String toString() {
+    final String type;
+    switch (units.getType()) {
+      case AA_ATTACKER:
+        type = "AA";
+        break;
+      case AA_DEFENDER:
+        type = "AD";
+        break;
+      case SUB_ATTACKER:
+        type = "SA";
+        break;
+      case SUB_DEFENDER:
+        type = "SD";
+        break;
+      case ATTACKER:
+        type = "AT";
+        break;
+      case DEFENDER:
+        type = "DE";
+        break;
+      default:
+        type = "??";
+        break;
+    }
+    final String out =
+        ""
+            + player.getName()
+            + " ("
+            + type
+            + ") "
+            + units.getAliveOrWaitingToDieFriendly()
+            + "("
+            + units.getAliveFriendly()
+            + ")"
+            + " vs "
+            + units.getAliveOrWaitingToDieEnemy()
+            + "("
+            + units.getAliveEnemy()
+            + ")"
+            + " has "
+            + String.format("%.4f", probability)
+            + "% chance"
+            + " w:"
+            + String.format("%.3f", winProbability)
+            + " l:"
+            + String.format("%.3f", loseProbability)
+            + " t:"
+            + String.format("%.3f", tieProbability)
+            + " b:"
+            + String.format("%.3f", badProbability)
+            + " [r:"
+            + round
+            + "]";
+
+    return out;
```

Comment:
2 Things:
1. No need to create a local variable out here, you can just return directly.
2. I thinkit's better to just wrap everything inside a single String.format

Reply:
Good point.  This function has changed a lot as I've added/removed things to help with debugging so it used to have additional variables and logic but those have been removed and I can simplify it further.